segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkedSegmentStorage.java

```diff
@@ -244,23 +239,42 @@ private void claimOwnership(MetadataTransaction txn, SegmentMetadata segmentMeta
                     segmentMetadata.getName(),
                     lastChunk.getName(),
                     lastChunk.getLength());
-            ChunkInfo chunkInfo = chunkStorage.getInfo(lastChunkName);
-            Preconditions.checkState(chunkInfo != null);
-            Preconditions.checkState(lastChunk != null);
-            // Adjust its length;
-            if (chunkInfo.getLength() != lastChunk.getLength()) {
-                Preconditions.checkState(chunkInfo.getLength() > lastChunk.getLength());
-                // Whatever length you see right now is the final "sealed" length of the last chunk.
-                lastChunk.setLength(chunkInfo.getLength());
-                segmentMetadata.setLength(segmentMetadata.getLastChunkStartOffset() + lastChunk.getLength());
-                txn.update(lastChunk);
-                log.debug("{} claimOwnership - Length of last chunk adjusted - segment={}, last chunk={}, Length={}.",
+            try {
+                ChunkInfo chunkInfo = chunkStorage.getInfo(lastChunkName);
+                Preconditions.checkState(chunkInfo != null);
+                Preconditions.checkState(lastChunk != null);
+                // Adjust its length;
+                if (chunkInfo.getLength() != lastChunk.getLength()) {
+                    Preconditions.checkState(chunkInfo.getLength() > lastChunk.getLength());
+                    // Whatever length you see right now is the final "sealed" length of the last chunk.
+                    lastChunk.setLength(chunkInfo.getLength());
+                    segmentMetadata.setLength(segmentMetadata.getLastChunkStartOffset() + lastChunk.getLength());
+                    txn.update(lastChunk);
+                    log.debug("{} claimOwnership - Length of last chunk adjusted - segment={}, last chunk={}, Length={}.",
+                            logPrefix,
+                            segmentMetadata.getName(),
+                            lastChunk.getName(),
+                            chunkInfo.getLength());
+                }
+            } catch (ChunkNotFoundException e) {
+                // This probably means that this instance is fenced out and newer instance truncated this segment.
+                // Try a commit of unmodified data to fail fast.
+                log.debug("{} claimOwnership - Last chunk was missing, failing fast - segment={}, last chunk={}.",
                         logPrefix,
                         segmentMetadata.getName(),
-                        lastChunk.getName(),
-                        chunkInfo.getLength());
+                        lastChunk.getName());
+                txn.update(segmentMetadata);
```

Comment:
I don't understand what this exception path is doing, why is it correct to update the metadata and commit given that we suspect that there is a new owner?

Reply:
During `openRead` or `openWrite` we try to reconcile the chunk length in metadata with the actual length on LTS.
There is a failover case where segment was last written by a segment store instance that is no more valid, and therefore needs ownership change. Then the current instance itself fails over to even newer segment store instance during the open calls. This most recent instance truncates the segment removing some of the chunks.
This causes open to fail with missing chunks exception. Therefore it does not attempt to write to metadata store and does not detect it has been fenced out. This error is "sticky" - it is tried again and again and we can't recover from it. Although it does not allow zombie instance to incorrectly write any new data, it still causes quite a few logs to get written. and sometimes it takes a while to detect failover.

The change basically forces instances to write to metadata store where it will immediately get exceptions related to being fenced out.

Reply:
ChunkNotFoundException in this case is only possible indication that , not confirmation. 

Missing chunk when metadata say it should exist is a pretty serious issue in normal circumstances and it should never ever happen. But it does happen in a very specific corner case when newer instance truncates segment without the zombie instance knowing about it. In this case ChunkNotFoundException shadows the underlying issue which is that this instance is no more valid - but because of the ChunkNotFoundException  it never commits and doesn't detect immediately.

This causes quite a bit spamming of logs until the instance detects that it is no longer a valid instance by some other means.

By trying to commit **unchanged** metadata, we force early detection of fencing out.
In case the metadata update actually succeeds then that means this instance is still the owner. In this case we simply throw the original exception as it might be a true issue. (The metadata update is a "no-op" as it doesn't change anything)

``` java
} catch (ChunkNotFoundException e) {
                // This probably means that this instance is fenced out and newer instance truncated this segment.
                // Try a commit of unmodified data to fail fast.
                log.debug("{} claimOwnership - Last chunk was missing, failing fast - segment={}, last chunk={}.",
                        logPrefix,
                        segmentMetadata.getName(),
                        lastChunk.getName());
                txn.update(segmentMetadata);
                txn.commit();
                throw e;
            }
```

Reply:
This is a convoluted way to disambiguate the scenario. My understanding is that:

1- If the `ChunkNotFoundException` is due to the instance being fenced out, then we expect `txn.update(segmentMetadata)` (or maybe `txn.commit()`) to throw.
2- If the `ChunkNotFoundException` is due to some other problem, then the original exception is rethrown, and the commit is essentially a no-op.

Assuming this is correct, is there a more direct way of executing (1)?

Reply:
In normal scenario, whenever an instance is fenced out then any further calls to `txn.commit()` throw `StorageMetadataWritesFencedOutException` .
However if we encounter any other problem/exception then code doesn't reach `txn.commit()`. This is how we want it and this is how it works. In this particular scenario also, the exactly the same is supposed to happen- attempt to change metadata should result in table store update (BK write) that is fenced out. 

However claimOwnership method is called from `openRead`/`openWrite` call which requires us to check the length of last chunk.

In corner case when new instance truncates first and zombie instance tries to open the same segment later, it finds the chunks missing and throws exception. This prevents it from committing (and therefore detecting fencing out in this case). 

Options -
1. Do not change - results in quite a bit spam in logs with ChunkNotFound exceptions. This will continue until zombie instance is able to contact ZK and detect that container assignment has changed. This takes time.

2. Attempt to fail early by committing  "no-op" change. If this instance is not fenced out the commit will succeed, in this case throw original exception.

I do not think there is an easier way to achieve this.

Reply:
The logic I think you are trying to implement is:

```
           } catch (ChunkNotFoundException e) {
                // This probably means that this instance is fenced out and newer instance truncated this segment.
                // Try a commit of unmodified data to fail fast.
                log.debug("{} claimOwnership - Last chunk was missing, failing fast - segment={}, last chunk={}.",
                        logPrefix,
                        segmentMetadata.getName(),
                        lastChunk.getName());
                if(amIFencedOut()) {
                    throw new StorageMetadataWritesFencedOutException();
                } else {
                    throw e;
                }
            }
```

Assuming this is right, you are implementing this if block using txn update and commit. My question is why you are not throwing the transaction you want directly rather than inducing via this no-op transaction. It clearly requires determining whether this instance is fenced out, is it not doable?

Reply:
Yes your code snippet is correct.

``` java
private void claimOwnership(MetadataTransaction txn, SegmentMetadata segmentMetadata) throws ChunkStorageException, StorageMetadataException {

        // Get the last chunk
        String lastChunkName = segmentMetadata.getLastChunk();
        if (null != lastChunkName) {
            ChunkMetadata lastChunk = (ChunkMetadata) txn.get(lastChunkName);
            try {
                ChunkInfo chunkInfo = chunkStorage.getInfo(lastChunkName);   ========> throws exception
                // Adjust its length;
                if (chunkInfo.getLength() != lastChunk.getLength()) {
                         // Adjust metadata  
                        //  Without this update part , we have serious problem of data loss/corruption.
                }
            } catch (ChunkNotFoundException e) {
               throw e;
           }

            // Claim ownership.
            // This is safe because the previous instance is definitely not an owner anymore. (even if this instance is no more owner)
          // If this instance is no more owner, then transaction commit will fail.So it is still safe.
           segmentMetadata.setOwnerEpoch(this.epoch);  
           segmentMetadata.setOwnershipChanged(true); ===> metadata update.

           // Update and commit
           // If This instance is fenced this update will fail.
           txn.commit();  =======>can't reach here because exception is thrown. 
    }
 }
```


1. If exception is thrown the normal txn.commit(); is not executed as that exception bubbles up.
2. We can't ignore ChunkNotFoundException and proceed to commit, because in case the commit succeeds then we have committed wrong/incomplete metadata and ignored a potentially serious error.

By committing no-change metadata - we are just forcing a write to BK that triggers StorageMetadataWritesFencedOutException being thrown.   If this commit succeeds then metadata is not incorrectly changed, and we then throw original exception .

I think the limitation here is that the only way to figure out `amIFencedOut()` is to actually commit something. 
We could also commit something useless data that is not used anywhere. But instead I am committing unchanged record.

Reply:
The level of indirection is hacky because there is really no need to have a transaction in this exception handling. If a new instance has taken over, then I understand that previous ledgers have been fenced and closed, in which case, we can check that directly with `BookKeeper#isClosed(ledgerId)` and there isn't a need to attempt an add entry:

https://bookkeeper.apache.org/docs/4.9.2/api/javadoc/org/apache/bookkeeper/client/BookKeeper.html#isClosed--

I understand that while the new instance is trying to take over, there is a period in which the ledger is recovering, and consequently the state is not closed. However, when the new instance takes over ownership and starts making changes to chunk metadata, the process of closing ledgers from former owners must have completed.

This is not a blocker, but I'd say we should consider fixing it and throwing the exception directly rather than going through a txn update and commit. If you agree, please create an issue to fix it. There might be other parts of the code that could benefit from such a change.

Reply:
Created tracking issue - https://github.com/pravega/pravega/issues/4987